File: src/corral/store/webhooks.py
"""Webhook configuration and delivery database operations."""

from __future__ import annotations

from datetime import datetime, timezone, timedelta
from typing import Any

from corral.store.connection import DatabaseManager


class WebhookStore(DatabaseManager):
    """Webhook configs and delivery queue CRUD operations."""

    async def _ensure_schema(self, conn) -> None:
        await conn.executescript("""
            CREATE TABLE IF NOT EXISTS webhook_configs (
                id                     INTEGER PRIMARY KEY AUTOINCREMENT,
                name                   TEXT NOT NULL,
                platform               TEXT NOT NULL,
                url                    TEXT NOT NULL,
                enabled                INTEGER NOT NULL DEFAULT 1,
                event_filter           TEXT NOT NULL DEFAULT '*',
                idle_threshold_seconds INTEGER NOT NULL DEFAULT 0,
                agent_filter           TEXT,
                low_confidence_only    INTEGER NOT NULL DEFAULT 0,
                consecutive_failures   INTEGER NOT NULL DEFAULT 0,
                created_at             TEXT NOT NULL,
                updated_at             TEXT NOT NULL
            );

            CREATE TABLE IF NOT EXISTS webhook_deliveries (
                id            INTEGER PRIMARY KEY AUTOINCREMENT,
                webhook_id    INTEGER NOT NULL,
                agent_name    TEXT NOT NULL,
                session_id    TEXT,
                event_type    TEXT NOT NULL,
                event_summary TEXT NOT NULL,
                status        TEXT NOT NULL DEFAULT 'pending',
                http_status   INTEGER,
                error_msg     TEXT,
                attempt_count INTEGER NOT NULL DEFAULT 0,
                next_retry_at TEXT,
                delivered_at  TEXT,
                created_at    TEXT NOT NULL
            );

            CREATE INDEX IF NOT EXISTS idx_webhook_deliveries_webhook
                ON webhook_deliveries(webhook_id, created_at DESC);

            CREATE INDEX IF NOT EXISTS idx_webhook_deliveries_pending
                ON webhook_deliveries(status, next_retry_at)
                WHERE status = 'pending';
        """)
        await conn.commit()

    # ── Webhook Configs ───────────────────────────────────────────────────

    async def list_webhook_configs(
        self, enabled_only: bool = False
    ) -> list[dict[str, Any]]:
        conn = await self._get_conn()
        sql = ("SELECT * FROM webhook_configs WHERE enabled = 1 ORDER BY created_at"
               if enabled_only else
               "SELECT * FROM webhook_configs ORDER BY created_at")
        rows = await (await conn.execute(sql)).fetchall()
        return [dict(r) for r in rows]

    async def get_webhook_config(self, webhook_id: int) -> dict[str, Any] | None:
        conn = await self._get_conn()
        row = await (await conn.execute(
            "SELECT * FROM webhook_configs WHERE id = ?", (webhook_id,)
        )).fetchone()
        return dict(row) if row else None

    async def create_webhook_config(
        self, name: str, platform: str, url: str,
        event_filter: str = "*", idle_threshold_seconds: int = 0,
        agent_filter: str | None = None, low_confidence_only: bool = False,
    ) -> dict[str, Any]:
        now = datetime.now(timezone.utc).isoformat()
        conn = await self._get_conn()
        cur = await conn.execute(
            """INSERT INTO webhook_configs
               (name, platform, url, enabled, event_filter,
                idle_threshold_seconds, agent_filter, low_confidence_only,
                consecutive_failures, created_at, updated_at)
               VALUES (?, ?, ?, 1, ?, ?, ?, ?, 0, ?, ?)""",
            (name, platform, url, event_filter, idle_threshold_seconds,
             agent_filter, int(low_confidence_only), now, now),
        )
        await conn.commit()
        return await self.get_webhook_config(cur.lastrowid)

    async def update_webhook_config(self, webhook_id: int, **fields) -> None:
        now = datetime.now(timezone.utc).isoformat()
        conn = await self._get_conn()
        allowed = {
            "name", "platform", "url", "enabled", "event_filter",
            "idle_threshold_seconds", "agent_filter", "low_confidence_only",
            "consecutive_failures",
        }
        set_clauses = ["updated_at = ?"]
        params: list[Any] = [now]
        for key, val in fields.items():
            if key in allowed:
                set_clauses.append(f"{key} = ?")
                params.append(val)
        params.append(webhook_id)
        await conn.execute(
            f"UPDATE webhook_configs SET {', '.join(set_clauses)} WHERE id = ?",
            params,
        )
        await conn.commit()

    async def delete_webhook_config(self, webhook_id: int) -> None:
        conn = await self._get_conn()
        # Manual cascade: delete deliveries first (no FK constraint in schema)
        await conn.execute(
            "DELETE FROM webhook_deliveries WHERE webhook_id = ?", (webhook_id,)
        )
        await conn.execute(
            "DELETE FROM webhook_configs WHERE id = ?", (webhook_id,)
        )
        await conn.commit()

    # ── Webhook Deliveries ────────────────────────────────────────────────

    async def create_webhook_delivery(
        self, webhook_id: int, agent_name: str, event_type: str,
        event_summary: str, session_id: str | None = None,
    ) -> dict[str, Any]:
        now = datetime.now(timezone.utc).isoformat()
        conn = await self._get_conn()
        cur = await conn.execute(
            """INSERT INTO webhook_deliveries
               (webhook_id, agent_name, session_id, event_type,
                event_summary, status, attempt_count, created_at)
               VALUES (?, ?, ?, ?, ?, 'pending', 0, ?)""",
            (webhook_id, agent_name, session_id, event_type, event_summary, now),
        )
        delivery_id = cur.lastrowid
        # Prune to 200 deliveries per webhook, excluding pending rows
        await conn.execute(
            "DELETE FROM webhook_deliveries WHERE webhook_id = ? "
            "AND status != 'pending' AND id NOT IN "
            "(SELECT id FROM webhook_deliveries WHERE webhook_id = ? "
            " AND status != 'pending' ORDER BY id DESC LIMIT 200)",
            (webhook_id, webhook_id),
        )
        await conn.commit()
        return {
            "id": delivery_id, "webhook_id": webhook_id,
            "status": "pending", "created_at": now,
            "agent_name": agent_name, "event_type": event_type,
            "event_summary": event_summary, "attempt_count": 0,
        }

    async def mark_webhook_delivery(
        self, delivery_id: int, status: str,
        http_status: int | None = None, error_msg: str | None = None,
        attempt_count: int | None = None, next_retry_at: str | None = None,
    ) -> None:
        now = datetime.now(timezone.utc).isoformat()
        conn = await self._get_conn()
        delivered_at = now if status == "delivered" else None
        await conn.execute(
            """UPDATE webhook_deliveries
               SET status = ?, http_status = ?, error_msg = ?,
                   attempt_count = COALESCE(?, attempt_count),
                   next_retry_at = ?, delivered_at = ?
               WHERE id = ?""",
            (status, http_status, error_msg, attempt_count,
             next_retry_at, delivered_at, delivery_id),
        )
        await conn.commit()

    async def get_pending_webhook_deliveries(
        self, limit: int = 50
    ) -> list[dict[str, Any]]:
        now = datetime.now(timezone.utc).isoformat()
        conn = await self._get_conn()
        rows = await (await conn.execute(
            """SELECT * FROM webhook_deliveries
               WHERE status = 'pending'
                 AND (next_retry_at IS NULL OR next_retry_at <= ?)
               ORDER BY created_at LIMIT ?""",
            (now, limit),
        )).fetchall()
        return [dict(r) for r in rows]

    async def list_webhook_deliveries(
        self, webhook_id: int, limit: int = 50
    ) -> list[dict[str, Any]]:
        conn = await self._get_conn()
        rows = await (await conn.execute(
            "SELECT * FROM webhook_deliveries WHERE webhook_id = ? "
            "ORDER BY created_at DESC LIMIT ?",
            (webhook_id, limit),
        )).fetchall()
        return [dict(r) for r in rows]

    async def get_last_event_times_by_agent(self) -> dict[str, str]:
        conn = await self._get_conn()
        rows = await (await conn.execute(
            "SELECT agent_name, MAX(created_at) as last_ts "
            "FROM agent_events GROUP BY agent_name"
        )).fetchall()
        return {r["agent_name"]: r["last_ts"] for r in rows if r["last_ts"]}

    async def idle_notification_exists(
        self, webhook_id: int, agent_name: str, threshold_seconds: int
    ) -> bool:
        cutoff = (
            datetime.now(timezone.utc) - timedelta(seconds=threshold_seconds)
        ).isoformat()
        conn = await self._get_conn()
        row = await (await conn.execute(
            "SELECT COUNT(*) as cnt FROM webhook_deliveries "
            "WHERE webhook_id = ? AND agent_name = ? AND event_type = 'idle' "
            "AND created_at >= ?",
            (webhook_id, agent_name, cutoff),
        )).fetchone()
        return bool(row and row["cnt"] > 0)

    async def increment_consecutive_failures(self, webhook_id: int) -> int:
        """Increment failure counter and return new value."""
        conn = await self._get_conn()
        await conn.execute(
            "UPDATE webhook_configs SET consecutive_failures = consecutive_failures + 1 "
            "WHERE id = ?", (webhook_id,),
        )
        await conn.commit()
        cfg = await self.get_webhook_config(webhook_id)
        return cfg["consecutive_failures"] if cfg else 0

    async def reset_consecutive_failures(self, webhook_id: int) -> None:
        conn = await self._get_conn()
        await conn.execute(
            "UPDATE webhook_configs SET consecutive_failures = 0 WHERE id = ?",
            (webhook_id,),
        )
        await conn.commit()

    async def auto_disable_webhook(self, webhook_id: int, reason: str) -> None:
        now = datetime.now(timezone.utc).isoformat()
        conn = await self._get_conn()
        await conn.execute(
            "UPDATE webhook_configs SET enabled = 0, updated_at = ? WHERE id = ?",
            (now, webhook_id),
        )
        await conn.commit()

File: coral-go/internal/store/webhooks.go
package store

import (
	"context"
	"database/sql"
	"fmt"
	"strings"
	"time"
)

// WebhookConfig represents a webhook subscription.
type WebhookConfig struct {
	ID                   int64   `db:"id" json:"id"`
	Name                 string  `db:"name" json:"name"`
	Platform             string  `db:"platform" json:"platform"`
	URL                  string  `db:"url" json:"url"`
	Enabled              int     `db:"enabled" json:"enabled"`
	EventFilter          string  `db:"event_filter" json:"event_filter"`
	IdleThresholdSeconds int     `db:"idle_threshold_seconds" json:"idle_threshold_seconds"`
	AgentFilter          *string `db:"agent_filter" json:"agent_filter"`
	LowConfidenceOnly    int     `db:"low_confidence_only" json:"low_confidence_only"`
	ConsecutiveFailures  int     `db:"consecutive_failures" json:"consecutive_failures"`
	CreatedAt            string  `db:"created_at" json:"created_at"`
	UpdatedAt            string  `db:"updated_at" json:"updated_at"`
}

// WebhookDelivery represents a webhook delivery attempt.
type WebhookDelivery struct {
	ID           int64   `db:"id" json:"id"`
	WebhookID    int64   `db:"webhook_id" json:"webhook_id"`
	AgentName    string  `db:"agent_name" json:"agent_name"`
	SessionID    *string `db:"session_id" json:"session_id,omitempty"`
	EventType    string  `db:"event_type" json:"event_type"`
	EventSummary string  `db:"event_summary" json:"event_summary"`
	Status       string  `db:"status" json:"status"`
	HTTPStatus   *int    `db:"http_status" json:"http_status,omitempty"`
	ErrorMsg     *string `db:"error_msg" json:"error_msg,omitempty"`
	AttemptCount int     `db:"attempt_count" json:"attempt_count"`
	NextRetryAt  *string `db:"next_retry_at" json:"next_retry_at,omitempty"`
	DeliveredAt  *string `db:"delivered_at" json:"delivered_at,omitempty"`
	CreatedAt    string  `db:"created_at" json:"created_at"`
}

// WebhookStore provides webhook config and delivery operations.
type WebhookStore struct {
	db *DB
}

// NewWebhookStore creates a new WebhookStore.
func NewWebhookStore(db *DB) *WebhookStore {
	return &WebhookStore{db: db}
}

// ── Webhook Configs ────────────────────────────────────────────────────

// ListWebhookConfigs returns all webhook configs.
func (s *WebhookStore) ListWebhookConfigs(ctx context.Context, enabledOnly bool) ([]WebhookConfig, error) {
	var configs []WebhookConfig
	query := "SELECT * FROM webhook_configs ORDER BY created_at"
	if enabledOnly {
		query = "SELECT * FROM webhook_configs WHERE enabled = 1 ORDER BY created_at"
	}
	err := s.db.SelectContext(ctx, &configs, query)
	return configs, err
}

// GetWebhookConfig returns a webhook config by ID.
func (s *WebhookStore) GetWebhookConfig(ctx context.Context, webhookID int64) (*WebhookConfig, error) {
	var cfg WebhookConfig
	err := s.db.GetContext(ctx, &cfg, "SELECT * FROM webhook_configs WHERE id = ?", webhookID)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	return &cfg, err
}

// CreateWebhookConfig creates a new webhook config.
func (s *WebhookStore) CreateWebhookConfig(ctx context.Context, name, platform, url string, agentFilter *string) (*WebhookConfig, error) {
	now := nowUTC()
	result, err := s.db.ExecContext(ctx,
		`INSERT INTO webhook_configs
		 (name, platform, url, enabled, event_filter, idle_threshold_seconds,
		  agent_filter, low_confidence_only, consecutive_failures, created_at, updated_at)
		 VALUES (?, ?, ?, 1, '*', 0, ?, 0, 0, ?, ?)`,
		name, platform, url, agentFilter, now, now)
	if err != nil {
		return nil, err
	}
	id, _ := result.LastInsertId()
	return s.GetWebhookConfig(ctx, id)
}

// UpdateWebhookConfig updates allowed fields on a webhook config.
func (s *WebhookStore) UpdateWebhookConfig(ctx context.Context, webhookID int64, fields map[string]interface{}) error {
	allowed := map[string]bool{
		"name": true, "platform": true, "url": true, "enabled": true,
		"event_filter": true, "idle_threshold_seconds": true,
		"agent_filter": true, "low_confidence_only": true, "consecutive_failures": true,
	}
	now := nowUTC()
	sets := []string{"updated_at = ?"}
	args := []interface{}{now}
	for k, v := range fields {
		if !allowed[k] {
			continue
		}
		sets = append(sets, fmt.Sprintf("%s = ?", k))
		args = append(args, v)
	}
	args = append(args, webhookID)
	_, err := s.db.ExecContext(ctx,
		fmt.Sprintf("UPDATE webhook_configs SET %s WHERE id = ?", strings.Join(sets, ", ")),
		args...)
	return err
}

// DeleteWebhookConfig deletes a webhook config and its deliveries.
func (s *WebhookStore) DeleteWebhookConfig(ctx context.Context, webhookID int64) error {
	tx, err := s.db.BeginTxx(ctx, nil)
	if err != nil {
		return err
	}
	defer tx.Rollback()

	tx.ExecContext(ctx, "DELETE FROM webhook_deliveries WHERE webhook_id = ?", webhookID)
	tx.ExecContext(ctx, "DELETE FROM webhook_configs WHERE id = ?", webhookID)
	return tx.Commit()
}

// ── Webhook Deliveries ─────────────────────────────────────────────────

// CreateWebhookDelivery creates a new pending delivery and prunes old ones.
func (s *WebhookStore) CreateWebhookDelivery(ctx context.Context, webhookID int64, agentName, eventType, eventSummary string, sessionID *string) (*WebhookDelivery, error) {
	now := nowUTC()
	result, err := s.db.ExecContext(ctx,
		`INSERT INTO webhook_deliveries
		 (webhook_id, agent_name, session_id, event_type, event_summary, status, attempt_count, created_at)
		 VALUES (?, ?, ?, ?, ?, 'pending', 0, ?)`,
		webhookID, agentName, sessionID, eventType, eventSummary, now)
	if err != nil {
		return nil, err
	}
	id, _ := result.LastInsertId()

	// Prune to 200 deliveries per webhook (excluding pending)
	s.db.ExecContext(ctx,
		`DELETE FROM webhook_deliveries WHERE webhook_id = ?
		 AND status != 'pending' AND id NOT IN
		 (SELECT id FROM webhook_deliveries WHERE webhook_id = ?
		  AND status != 'pending' ORDER BY id DESC LIMIT 200)`,
		webhookID, webhookID)

	return &WebhookDelivery{
		ID: id, WebhookID: webhookID, AgentName: agentName,
		SessionID: sessionID, EventType: eventType, EventSummary: eventSummary,
		Status: "pending", AttemptCount: 0, CreatedAt: now,
	}, nil
}

// MarkWebhookDelivery updates the status of a delivery.
func (s *WebhookStore) MarkWebhookDelivery(ctx context.Context, deliveryID int64, status string, httpStatus *int, errorMsg *string, attemptCount *int, nextRetryAt *string) error {
	now := nowUTC()
	var deliveredAt *string
	if status == "delivered" {
		deliveredAt = &now
	}
	_, err := s.db.ExecContext(ctx,
		`UPDATE webhook_deliveries
		 SET status = ?, http_status = ?, error_msg = ?,
		     attempt_count = COALESCE(?, attempt_count),
		     next_retry_at = ?, delivered_at = ?
		 WHERE id = ?`,
		status, httpStatus, errorMsg, attemptCount, nextRetryAt, deliveredAt, deliveryID)
	return err
}

// GetPendingWebhookDeliveries returns deliveries ready for sending.
func (s *WebhookStore) GetPendingWebhookDeliveries(ctx context.Context, limit int) ([]WebhookDelivery, error) {
	now := time.Now().UTC().Format(time.RFC3339)
	var deliveries []WebhookDelivery
	err := s.db.SelectContext(ctx, &deliveries,
		`SELECT * FROM webhook_deliveries
		 WHERE status = 'pending' AND (next_retry_at IS NULL OR next_retry_at <= ?)
		 ORDER BY created_at LIMIT ?`,
		now, limit)
	return deliveries, err
}

// ListWebhookDeliveries returns recent deliveries for a webhook.
func (s *WebhookStore) ListWebhookDeliveries(ctx context.Context, webhookID int64, limit int) ([]WebhookDelivery, error) {
	var deliveries []WebhookDelivery
	err := s.db.SelectContext(ctx, &deliveries,
		`SELECT * FROM webhook_deliveries WHERE webhook_id = ?
		 ORDER BY created_at DESC LIMIT ?`,
		webhookID, limit)
	return deliveries, err
}

// IncrementConsecutiveFailures increments the failure counter and returns the new value.
func (s *WebhookStore) IncrementConsecutiveFailures(ctx context.Context, webhookID int64) (int, error) {
	_, err := s.db.ExecContext(ctx,
		"UPDATE webhook_configs SET consecutive_failures = consecutive_failures + 1 WHERE id = ?",
		webhookID)
	if err != nil {
		return 0, err
	}
	cfg, err := s.GetWebhookConfig(ctx, webhookID)
	if err != nil || cfg == nil {
		return 0, err
	}
	return cfg.ConsecutiveFailures, nil
}

// ResetConsecutiveFailures resets the failure counter to zero.
func (s *WebhookStore) ResetConsecutiveFailures(ctx context.Context, webhookID int64) error {
	_, err := s.db.ExecContext(ctx,
		"UPDATE webhook_configs SET consecutive_failures = 0 WHERE id = ?", webhookID)
	return err
}

// AutoDisableWebhook disables a webhook due to too many failures.
func (s *WebhookStore) AutoDisableWebhook(ctx context.Context, webhookID int64) error {
	now := nowUTC()
	_, err := s.db.ExecContext(ctx,
		"UPDATE webhook_configs SET enabled = 0, updated_at = ? WHERE id = ?",
		now, webhookID)
	return err
}
